### PR TITLE
Blog post on communicating software development

### LINKED ISSUE #29: Comparison of lifecycle frameworks
- RECON
- `r-lib/lifecycle`
- repostatus
- package numbering

### PR DESCRIPTION
**Pull request by joshwlambert**
New blog post on the methods to communicate the development status of open-source software, such as the Epiverse-TRACE R packages.

This PR closes #29.

- [x] The post specifies a license if you don't want to use the default CC BY
- [x] All authors have an ORCID iD
- [x] Relevant keywords / tags has been added. In particular, if you want your post to be shared on R-bloggers, you must tag it with `R`
- [x] Images or other external resources have been committed and pushed
- [x] The post uses pure quarto syntax, rather than HTML or R code, unless necessary

Right before merging:

- [x] The `date` field has been updated
- [x] A PR has been opened in the [`blueprints`](https://github.com/epiverse-trace/blueprints) to link to this post
- [x] The post has been re-rendered and content of the `_freeze/` folder is up-to-date

### CODE REVIEW

File: posts/comm-software-devel/index.qmd [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+
+Overall semantic versioning provides what they describe as a "formal specification" to facilitate management of package development and the dependencies of that package. It is the most widely-used versioning system and therefore will be understood by a wide variety of users and developers.
+
+Some of the critique raised for semantic versioning is the difficulty of defining how changes correspond to a version increment. Semantic versioning states only breaking changes warrant major releases, but a vast re-write of a code base may also justify a major version change. Different breaking changes have different magnitudes, therefore a change to a single exported function or a change to every exported function will be communicated in a single, equal, version increment.
+
+### Alternatives to seminatic versioning
```
Comment (Bisaloo) [inline posts/comm-software-devel/index.qmd:35]:
```suggestion
### Alternatives to semiantic versioning
```

File: posts/comm-software-devel/index.qmd [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+  - name: "Joshua W. Lambert"
+    orcid: "0000-0001-5218-3046"
+date: "2023-07-12"
+categories: [open-source, R, R package, versioning, lifecycles, community]
+format:
+  html:
+    toc: true
+---
+
+Software is not immediately stable when being developed. It undergoes design changes, changes to the user interface (application programming interface, API), and features get added or removed over time. Software in a open-source setting, in which the code is publicly hosted in a repository (e.g., Github, GitLab, Bitbucket), allows anyone to track developments. It also allows the developer community to easily contribute to the software.
```
[the remark this answers is not in the document]
Comment (joshwlambert) [reply posts/comm-software-devel/index.qmd:13]: Thanks for bringing this up. I hadn't previously thought about it in terms of the working method. I think outside of developers these concepts are not very familiar and therefore might require some explanation (although most of the readership may be developers). Therefore, I'll leave these concepts out as I don't want to increase the size of the post. 

However, if we plan to do a follow up post (including Epiverse-TRACE group input) I think discussing these ideas, and how they influence versioning and breaking changes, would be good.

File: posts/comm-software-devel/index.qmd [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+One such philosophy is [semantic versioning](https://semver.org/) (SemVer). This aims to describe the stage of software development by attaching semantics (i.e. meaning) to the format and numbering of versions. The version system works through three numbers, each separated by a dot. The numbers, from left to right, convey major version, minor version and patch version. As an example, 0.5.2, is newer than 0.3.9.
+
+Employing semantic versioning in ones code development allows others to determine whether a package has undergone substantial development and testing, and informs to whether it would make a suitable package to use in a script or as a dependency for another package. Semantic versioning also describes the changes made to a package. As explained on [their website](https://semver.org/), incrementing the major version implies a breaking change, a minor increment is a backwards compatible change and lastly patches are mostly applied to bug fixes. This aids users in understanding whether they should continue using a package, whether their package needs updating due to a breaking change or whether they need to install the newest version because a bug was recently fixed.
```
[the remark this answers is not in the document]
Comment (joshwlambert) [reply posts/comm-software-devel/index.qmd:27]: Sorry I don't think I understand your point. Should I link to package release notes (or blog posts in the case of tidyverse) that discuss releases? I'm guessing there won't be any for patch releases. Or to link to packages that are currently in each stage, e.g., 2.1.0, 0.6.0 & 0.0.1?

File: posts/comm-software-devel/index.qmd [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+
+In R there are several packages that work with versioning, and specifically semantic versioning. The {semver} package provides functions for parsing, rendering and comparing versions. There is also the {semverutils} R package which provides similar functionality using `R6`. The {usethis} package provides handy utility functions for changing the versions of R packages (`usethis::use_version()` and `usethis::use_dev_version()`). R also comes with a `package_version()` function for creating and validating versions.
```
[the remark this answers is not in the document]
Comment (joshwlambert) [reply posts/comm-software-devel/index.qmd]: Resolved in 48abe426d0bde75c27f1e420eeedd597166207b2

File: posts/comm-software-devel/index.qmd (hunk 1 above)
Comment (joshwlambert) [reply posts/comm-software-devel/index.qmd]: Resolved in 6cd144edfa56aba433a5eff3a968d895b144bbe1

File: posts/comm-software-devel/index.qmd [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+
+### Alternatives to seminatic versioning
+
+There are several other versioning frameworks aside from semantic versioning. One common option is [calendar versioning](https://calver.org/) (CalVer). The format of CalVer is usually year-month (YY-MM), or year-month-day (YY-MM-DD), depending on the regularity of releases, and allows appending tags (micros or modifiers, e.g. YY-MM.1).
+
+Other versioning schemes can appear similar to semantic versioning, but do not follow the guidelines around version semantics. In these cases, a bump in the major version may not relate to a breaking change. Additionally, other numbers can be attached to the traditional x.y.z format, such as build numbers. Build number versioning adds an extra number to specify the build (x.y.z.build_number). There are many other variants but covering all versioning systems is outside the scope of this post.
+
+### Versioning an R package
+
+There are some restrictions on valid version numbers for R packages. The [guidelines state](https://cran.r-project.org/doc/manuals/R-exts.html):
```
[the remark this answers is not in the document]
Comment (joshwlambert) [reply posts/comm-software-devel/index.qmd]: Resolved in 6cd144edfa56aba433a5eff3a968d895b144bbe1

File: posts/comm-software-devel/index.qmd [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+### Reconverse
+
+Reconverse provides four stages of software development:
+
+1. concept
+2. experimental
+3. maturing
+4. stable
+
+A difference between {lifecycles} and reconverse is the explicit connection between semantic versioning and development stage in [reconverse](https://www.reconverse.org/lifecycle.html). The transitions between experimental, maturing and stable are linked to the versioning less than 0.1.0, less than 1.0.0 and greater than 1.0.0, respectively.
```
[the remark this answers is not in the document]
Comment (joshwlambert) [reply posts/comm-software-devel/index.qmd]: Resolved in 6cd144edfa56aba433a5eff3a968d895b144bbe1

File: posts/comm-software-devel/index.qmd (hunk 3 above)
[the remark this answers is not in the document]
Comment (Bisaloo) [reply posts/comm-software-devel/index.qmd:27]: > Should I link to package release notes (or blog posts in the case of tidyverse) that discuss releases? I'm guessing there won't be any for patch releases.

Yes, this. To give a feel of what kind of changes might lead to each kind of releases. You won't find blog posts for patch releases but they still have release notes in many cases.

[the remark this answers is not in the document]
Comment (joshwlambert) [reply posts/comm-software-devel/index.qmd:27]: Resolved in 5b87fd50e0cf1ce8c5a62dcd26977e933ed42160

**Comment by joshwlambert** [comment]
@Bisaloo Thanks for the comments! 

I've updated the date on the post for tomorrow. Let me know if you'd prefer to merge the PR then or if you're happy for me to do it.

> do you want to discuss the task you volunteered to lead ("Identify lifecycle stage from which our packages can be taken as dependencies") in this post or somewhere else?

I will leave this for a potential subsequent post as I don't want to increase the size of this post.

**Comment by pratikunterwegs** [comment]
Looks good to me, I thought it was well written and easy to understand. You've already got feedback that should be sufficient for the post.

**Pull request merged**